docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,262 @@
+
+This is both dangerous as the request is not yet validated but is changing state of the twin as well
+as not required as at the end of processing the message we do sync changes on the twin.
+
+We should instead:
+
+- Clear the cache in the Function as we do currently.
+- Function returns if we are the losing LNS. If we are the losing one, we drop the message here,
+  mark ourselves as the losing gateway etc.
+- Otherwise we process message normally.
+- And at the end force update the twin.
+
+Alternative considered but discarded: use the Function to do both operations
+
+- update the device twin frame counter to 0 (it needs a DeviceClient that would
+  cause a temporary connection switch)
+- Clear or update the cache entry with frame counter down and up to 0.
+- Returns the result to the LNS: whether it was the winning or losing one
+- LNS reacts as described in the [main data flow section](#main-data-message-flow)
+
+### Main data message flow
+
+Assuming the topology:
+
+```mermaid
+flowchart LR;
+    Device-->LBS1-->LNS1-->Function;
+    Device-->LBS2-->LNS2-->Function;
+    LNS1-->IoTHub;
+    LNS2-->IoTHub;
+```
+
+where Device sends data message A and then B.
+
+Here is a rundown of what should happen marked in **bold**:
+
+1. Device sends first data message A.
+1. We assume that LNS1 gets the message first. **LNS1 checks against
+   an in-memory dictionary DevEui -> flag** and since it has not seen this DevEui before (flag is
+   false) contacts the GetDevice.
+1. The Function hasn't seen this DevEui either and therefore does not have an assigned LNS for it
+  yet. LNS1 wins the race and gets immediately a response and processes the message upstream.
+1. LNS2 eventually receives message A, **checks its local dictionary** and also contacts the FunctionBundler since it does not have prior info about this DevEui.
+1. The Function responds to LNS2 that it lost the race to process this message.
+1. Since deduplication strategy is Drop, LNS2 drops the message immediately, therefore no
+   connection to Iot Hub is opened and only LNS1 has the connection to Iot Hub. **LNS2 also notes in
+   memory that it was the losing gateway for this DevEui**.
+1. When message B gets send (with a higher frame counter), assuming that this time LNS2 gets it
+   first it **checks again its local dictionary it's not the preferred LNS for this device and
+   therefore delays itself X ms before contacting the FunctionBundler**.
+   - Here we can *not* simply drop as LNS1 might not be available anymore (due to a crash, device not
+     in range etc).
+1. This delay gives LNS1 a time advantage to reach the FunctionBundler first and win the race again, failing
+    back to the previous case of message A. The active connection stays with LNS1.
+1. If this delay is not sufficient for LNS1 to win the race, LNS2 will contact the FunctionBundler which
+   now awards LNS2 as the "winning" LNS. LNS2 will process message upstream (therefore the active
```

Comment:
I think here a step is missing:

LNS2 contacts the IoT Hub to refresh its device twins. 

As the cache entry might be stale?